nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/marshallers/csv/DocumentModelCSVWriter.java

```diff
@@ -0,0 +1,137 @@
+    }
+
+    @Inject
+    protected SchemaManager schemaManager;
+
+    @Inject
+    protected MarshallerRegistry registry;
+
+    @Override
+    public void write(DocumentModel entity, Class<?> clazz, Type genericType, MediaType mediatype, OutputStream out)
+            throws IOException {
+        writeSystem(out, entity);
+        // schemas are sorted for reproducibility
+        List<String> schemas = new ArrayList<>(Arrays.asList(entity.getSchemas()));
+        Collections.sort(schemas);
+        for (int i = 0; i < schemas.size(); i++) {
+            writeSchema(out, entity, schemas.get(i));
+            if (i < schemas.size() - 1) {
+                writeSeparator(out);
+            }
+        }
+    }
+
+    /**
+     * CSV must have consistent column number, null and empy values have to be written.
+     */
+    protected void writeSystem(OutputStream out, DocumentModel doc) throws IOException {
+        writeWithSeparator(out, doc.getId());
+        writeWithSeparator(out, doc.getPathAsString());
+        writeWithSeparator(out, doc.getRepositoryName());
+        writeWithSeparator(out, doc.getType());
+        writeWithSeparator(out, doc.getRef() != null ? doc.getCurrentLifeCycleState() : null);
+        writeWithSeparator(out, doc.getTitle());
+        writeWithSeparator(out, doc.getParentRef() != null ? doc.getParentRef().toString() : null);
+        writeWithSeparator(out, Boolean.toString(doc.isTrashed()));
+        writeWithSeparator(out, Boolean.toString(doc.isVersion()));
+        writeWithSeparator(out, Boolean.toString(doc.isProxy()));
+        writeWithSeparator(out, Boolean.toString(doc.isCheckedOut()));
+        writeWithSeparator(out, doc.getChangeToken());
+        writeWithSeparator(out, doc.getVersionLabel());
+        Lock lock = doc.getLockInfo();
+        if (lock != null) {
+            writeWithSeparator(out, lock.getOwner());
+            writeWithSeparator(out, ISODateTimeFormat.dateTime().print(new DateTime(lock.getCreated())));
+        } else {
+            writeWithSeparator(out, null);
+            writeWithSeparator(out, null);
+        }
+        if (doc.hasSchema("dublincore")) {
+            Calendar cal = (Calendar) doc.getPropertyValue("dc:modified");
+            if (cal != null) {
+                writeWithSeparator(out, DateParser.formatW3CDateTime(cal.getTime()));
+            }
+        } else {
+            writeWithSeparator(out, null);
+        }
+    }
+
+    protected void writeSchema(OutputStream out, DocumentModel entity, String schemaName) throws IOException {
+        Writer<Property> propertyWriter = registry.getWriter(null, Property.class, TEXT_CSV_TYPE);
+        // provides the current document to the property writer
+        Schema schema = schemaManager.getSchema(schemaName);
+        List<Field> fields = new ArrayList<>(schema.getFields());
+        Collections.sort(fields, (o1, o2) -> o1.getName().getLocalName().compareTo(o2.getName().getLocalName()));
```

Comment:
In case you have to give a `Comparator`, the way below is preferred (less overhead, import, etc):
```
fields.sort((o1, o2) -> o1.getName().getLocalName().compareTo(o2.getName().getLocalName()));
```

Reply:
Yes although it's even better to abstract the comparator itself:

    fields.sort(Comparator.comparing(o -> o.getName().getLocalName()));